src/logger/src/metrics.rs

```diff
@@ -251,48 +251,45 @@ pub struct SharedIncMetric(AtomicUsize, AtomicUsize);
 pub struct SharedStoreMetric(AtomicUsize);
 
 impl IncMetric for SharedIncMetric {
-    // While the order specified for this operation is still Relaxed, the actual instruction will
-    // be an asm "LOCK; something" and thus atomic across multiple threads, simply because of the
-    // fetch_and_add (as opposed to "store(load() + 1)") implementation for atomics.
-    // TODO: would a stronger ordering make a difference here?
     fn add(&self, value: usize) {
-        self.0.fetch_add(value, Ordering::Relaxed);
+        self.0.fetch_add(value, Ordering::AcqRel);
     }
 
     fn count(&self) -> usize {
-        self.0.load(Ordering::Relaxed)
+        self.0.load(Ordering::Acquire)
     }
 }
 
 impl StoreMetric for SharedStoreMetric {
     fn fetch(&self) -> usize {
-        self.0.load(Ordering::Relaxed)
+        self.0.load(Ordering::Acquire)
     }
 
     fn store(&self, value: usize) {
-        self.0.store(value, Ordering::Relaxed);
+        self.0.store(value, Ordering::Release);
     }
 }
 
 impl Serialize for SharedIncMetric {
     /// Reset counters of each metrics. Here we suppose that Serialize's goal is to help with the
     /// flushing of metrics.
     /// !!! Any print of the metrics will also reset them. Use with caution !!!
+    /// !!! This function is not thread safe !!!
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         // There's no serializer.serialize_usize() for some reason :(
-        let snapshot = self.0.load(Ordering::Relaxed);
-        let res = serializer.serialize_u64(snapshot as u64 - self.1.load(Ordering::Relaxed) as u64);
+        let snapshot = self.count();
+        let res = serializer.serialize_u64(snapshot as u64 - self.1.load(Ordering::Acquire) as u64);
 
         if res.is_ok() {
-            self.1.store(snapshot, Ordering::Relaxed);
+            self.1.store(snapshot, Ordering::Release);
```

Comment:
I'm not sure if this is correct. If we can serialize from multiple threads, I think `snapshot` can be different on each one and the value that ends up being assigned to `self.1` in the end (`self.1.store(snapshot, Ordering::Release);`) won't necessarily be the most recent one. So we can end up in the situation where self.0 < self.1.

A solution would be to have another atomic bool that is marking whether there is a serialization in progress, and if there is, don't do anything. Although this wouldn't guarantee the serialization of the latest values.

Reply:
`self.0` is incremented atomically from any thread. If we have Relaxed semantics, it may have different values on multiple threads when serialising.
If using Acquire semantics, like we do in `self.count()`, we can be sure that any writes to `self.0` with `Release` semantics (like we do in add()) are going to be executed and visible before the load.

I believe that the scenario you are describing is possible prior to this PR, due to the Relaxed ordering.

> A solution would be to have another atomic bool that is marking whether there is a serialization in progress, and if there is, don't do anything. Although this wouldn't guarantee the serialization of the latest values.

This is essentially simulating a mutex

Reply:
anyway, I believe that the usage of the metrics system in signal handlers needs a redesign with thread-safety in mind.
Since this second commit is pretty controversial and may be introducing some performance degradation on ARM, I will remove it from this PR to unblock it and I will open a new issue for it

Reply:
Would something like the following be possible ?
```
+-----------------------------------+-----------------------------------+
| Thread 1                          | Thread 2                          |
+-----------------------------------+-----------------------------------+
| let snapshot = self.count(); -> 5 | ...                               |
+-----------------------------------+-----------------------------------+
| ...                               | inc metric -> 6                   |
+-----------------------------------+-----------------------------------+
| ...                               | let snapshot = self.count(); -> 6 |
+-----------------------------------+-----------------------------------+
| ...                               | store snapshot in self.1 -> 6     |
+-----------------------------------+-----------------------------------+
| store snapshot in self.1 -> 5     |                                   |
+-----------------------------------+-----------------------------------+
```

Reply:
Actually this is not possible since we only write the metrics from the VMM thread or the signal handler.
In the signal handler we now only use StoreMetrics so this would not be possible

Reply:
The only known issue with the current form of this PR is that the other (non-signal) SharedIncMetrics may be slightly outdated when written from the signal handler. But I think this is a fine compromise since the process is going to be killed anyway

Reply:
Ok, makes sense. Anyway this is hard to reason about and we need to make sure that `IncMetrics` won't be added to the signal handler thread. So should we mark this method as unsafe since it's not thread safe ? Or what would be the best way to do this ?

Reply:
I can't think of a good way of preventing this, except for code reviews.
Rust's unsafe keyword is not a good fit in this case because it does not express thread-unsafety.
For example `Rc` is not thread safe but is not marked as unsafe.

The one way of showing that a type is not thread safe is by not implementing the `Sync` trait, but this is required since the type is a `lazy_static`.

This race condition is furthermore detailed in the comment above `Metrics.write()` function. Do you have any other ideas or do you think this is sufficient?

Reply:
It doesn't seem sufficient to me because if I had to add a new metric I can imagine that it would be very easy to miss these comments. I wouldn't check the comments for the `Metrics.write()` or `serialize()` methods. It would be the same even if the `serialize()` method would be unsafe.

But I don't have a solution. I'm still thinking.

Reply:
If the type is not thread safe than `Sync` should not be implemented for it. Otherwise we're breaking the Rust safety principles and Rust cannot protect and benefit of the "fearless concurrency" anymore. The problem seems to be that we're doing more things than we should be doing in a context of the signal handler, so maybe a solution would be to simplify the signal handler. This will come at the expense of maybe metrics being missed, but it is arguably better than opening potentially other safety problems.

Reply:
We have discussed removing logging and metrics systems from the signal handlers.
One idea is to use specialised files for dumping abrupt exit information from signal and panic handlers (much like a core dump).

This is a separate feature and will be worked on in the future. It requires careful thought and more investment.

This PR addresses the issue surgically, since it is creating much trouble for us with intermittent test failures.
We are not introducing problems with this PR, we are applying a patch over the known occurrence of this problem.

Reply:
Created an issue for it: https://github.com/firecracker-microvm/firecracker/issues/2899

Reply:
> It doesn't seem sufficient to me because if I had to add a new metric I can imagine that it would be very easy to miss these comments. I wouldn't check the comments for the Metrics.write() or serialize() methods. It would be the same even if the serialize() method would be unsafe.

I realised that the signal handler generation macro is hardcoded to use the `store` function, which is only defined for `StoreMetrics`. It would require a separate function in order to add an IncMetric to a signal handler, like it is for SIGPIPE. This should be even more evident in code reviews